add migration guide for sdk 0.14 (#2516)
operator-framework/operator-sdk: doc/migration/version-upgrade-guide.md

### Go version

> 

Reviewer: Remove the "Go version" header. Ditto for other v0.14.x.

	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.0.0-20191016115521-756ffa5af0bd
	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.0.0-20191016112429-9587704a8ad4
	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.0.0-20191016114939-2b2b218dc1df
	k8s.io/kube-proxy => k8s.io/kube-proxy v0.0.0-20191016114407-2e83b6f20229
	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.0.0-20191016114748-65049c67a58b
	k8s.io/kubectl => k8s.io/kubectl v0.0.0-20191016120415-2ed914427d51
	k8s.io/kubelet => k8s.io/kubelet v0.0.0-20191016114556-7841ed97f1b2
	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.0.0-20191016115753-cf0698c3a16b
	k8s.io/metrics => k8s.io/metrics v0.0.0-20191016113814-3b1a734dba6e
	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
)
replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309 // Required by Helm
```

- Run `go mod tidy` to update the project modules
- Run the command `operator-sdk generate k8s` to ensure that your resources will be updated
- Run the command `operator-sdk generate crds` to regenerate CRDs

### Deprecations

> Note that `github.com/operator-framework/operator-sdk/pkg/restmapper` was deprecated in favor of the `DynamicRESTMapper` implementation in [controller-runtime](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client/apiutil#NewDiscoveryRESTMapper). Users should migrate to controller-runtime's implementation, which is a drop-in replacement.

Reviewer: ```suggestion
The `github.com/operator-framework/operator-sdk/pkg/restmapper` package was deprecated in favor of the `DynamicRESTMapper` implementation in [controller-runtime](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client/apiutil#NewDiscoveryRESTMapper). Users should migrate to controller-runtime's implementation, which is a drop-in replacement.
```

Replace:
```
github.com/operator-framework/operator-sdk/pkg/restmapper.DynamicRESTMapper
```

With:

```
sigs.k8s.io/controller-runtime/pkg/client/apiutil.DynamicRESTMapper
```

> ### Break Changes

Reviewer: ```suggestion
### Breaking Changes
```

#### Add `operator_sdk.util` Ansible collection

Also, the Ansible module `k8s_status` was extracted and it is now providing the `operator_sdk.util` Ansible collection. See [developer_guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/developer_guide.md#custom-resource-status-management) for new usage. Update the end of the `main.yaml` file with the `collections` after the `dependencies` block: 

```yaml
...
dependencies: []
  # List your role dependencies here, one per line. Be sure to remove the '[]' above,
  # if you add dependencies to this list.`
  # if you add dependencies to this list.
collections:
- operator_sdk.util
...

> ```

Reviewer: `````suggestion
The Ansible module `k8s_status` was extracted and is now provided by the `operator_sdk.util` Ansible collection. See [developer_guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/developer_guide.md#custom-resource-status-management) for new usage.

To use the collection in a role, declare it at the root level in `meta/main.yaml`:

```
collections:
- operator_sdk.util
```

To use it in a playbook, declare it in the play:
```
- hosts: all
  collections:
   - operator_sdk.util
  tasks:
   - k8s_status:
       api_version: app.example.com/v1
       kind: Foo
       name: "{{ meta.name }}"
       namespace: "{{ meta.namespace }}"
       status:
         foo: bar
```

You can also use the fully-qualified name without declaring the collection:
```
   - operator_sdk.util.k8s_status:
       api_version: app.example.com/v1
       kind: Foo
       name: "{{ meta.name }}"
       namespace: "{{ meta.namespace }}"
       status:
         foo: bar
```
`````

### Notable Changes

These notable changes contain just the most important user-facing changes. See the [CHANGELOG](https://github.com/operator-framework/operator-sdk/blob/master/CHANGELOG.md#v0141) for details of the release.

#### Ansible version update

> Note that the Ansible version in the init projects from `2.6` to `2.9` for collections support. Update the `main.yaml` file :

Reviewer: ```suggestion
The Ansible version in the init projects was upgraded from `2.6` to `2.9` for collections support. Update the `meta/main.yaml` file.
```

	k8s.io/metrics => k8s.io/metrics v0.0.0-20191016113814-3b1a734dba6e
	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
)
replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309 // Required by Helm
```

- Run `go mod tidy` to update the project modules
- Run the command `operator-sdk generate k8s` to ensure that your resources will be updated
- Run the command `operator-sdk generate crds` to regenerate CRDs

### Deprecations

The `github.com/operator-framework/operator-sdk/pkg/restmapper` package was deprecated in favor of the `DynamicRESTMapper` implementation in [controller-runtime](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client/apiutil#NewDiscoveryRESTMapper). Users should migrate to controller-runtime's implementation, which is a drop-in replacement.

Replace:
```
github.com/operator-framework/operator-sdk/pkg/restmapper.DynamicRESTMapper
```

With:

```
sigs.k8s.io/controller-runtime/pkg/client/apiutil.DynamicRESTMapper
```

### Breaking Changes

#### Add `operator_sdk.util` Ansible collection

> Also, the Ansible module `k8s_status` was extracted and it is now providing the `operator_sdk.util` Ansible collection. See [developer_guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/developer_guide.md#custom-resource-status-management) for new usage. Update the end of the `main.yaml` file with the `collections` after the `dependencies` block: 

Reviewer: Remove the above line.

> Also, the Ansible module `k8s_status` was extracted and it is now providing the `operator_sdk.util` Ansible collection. See [developer_guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/developer_guide.md#custom-resource-status-management) for new usage. Update the end of the `main.yaml` file with the `collections` after the `dependencies` block: 

Author: ```suggestion
```

	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.0.0-20191016115521-756ffa5af0bd
	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.0.0-20191016112429-9587704a8ad4
	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.0.0-20191016114939-2b2b218dc1df
	k8s.io/kube-proxy => k8s.io/kube-proxy v0.0.0-20191016114407-2e83b6f20229
	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.0.0-20191016114748-65049c67a58b
	k8s.io/kubectl => k8s.io/kubectl v0.0.0-20191016120415-2ed914427d51
	k8s.io/kubelet => k8s.io/kubelet v0.0.0-20191016114556-7841ed97f1b2
	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.0.0-20191016115753-cf0698c3a16b
	k8s.io/metrics => k8s.io/metrics v0.0.0-20191016113814-3b1a734dba6e
	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
)
replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309 // Required by Helm
```

- Run `go mod tidy` to update the project modules
- Run the command `operator-sdk generate k8s` to ensure that your resources will be updated
- Run the command `operator-sdk generate crds` to regenerate CRDs

> ### Deprecations on SDK

Reviewer: ```suggestion
### Deprecations
```

Note that `github.com/operator-framework/operator-sdk/pkg/restmapper` was deprecated in favor of the `DynamicRESTMapper` implementation in [controller-runtime](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client/apiutil#NewDiscoveryRESTMapper). Users should migrate to controller-runtime's implementation, which is a drop-in replacement.

Replace:
```
github.com/operator-framework/operator-sdk/pkg/restmapper.DynamicRESTMapper
```

With:

```
sigs.k8s.io/controller-runtime/pkg/client/apiutil.DynamicRESTMapper
```

### Notable Changes

> **(Valid for Ansible based-operators only)** 

Reviewer: Looks like you're mixing bold text and headers. This needs to be consistent and have the proper grouping (i.e. Ansible-related notes under an Ansible header and Helm-related notes under a Helm header)

For example

```md
### Notable Changes

#### Ansible Operator

... Ansible stuff here ...

#### Helm Operator

... Helm stuff here ...
```

> Note that the Ansible version in the init projects from `2.6` to `2.9` for collections support. Also, the Ansible module `k8s_status` was extracted and it is now providing the `operator_sdk.util` Ansible collection. See [developer_guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/developer_guide.md#custom-resource-status-management) for new usage.

Reviewer: ```suggestion
Note that the Ansible version in the init projects was upgraded from `2.6` to `2.9` for collections support. Also, the Ansible module `k8s_status` was extracted and is now provided by the `operator_sdk.util` Ansible collection. See [developer_guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/developer_guide.md#custom-resource-status-management) for new usage.
```

Reviewer: > Also, the Ansible module `k8s_status` was extracted and is now provided by the `operator_sdk.util` Ansible collection.

This is a breaking change and should be called out separately.

> #### (Optional) Update the main.yaml file

Reviewer: Given the above breaking change, is this actually optional?

In order to get the above changes in your ansble based-operator project. 

Replace:
```yaml
...
 min_ansible_version: 2.6
...
```

With: 
```yaml
...
 min_ansible_version: 2.9
...
```

And then, add at the `collections` at the end of the file after the `dependencies` block. 

Example:
```yaml
...
dependencies: []
  # List your role dependencies here, one per line. Be sure to remove the '[]' above,
  # if you add dependencies to this list.`
  # if you add dependencies to this list.
collections:
- operator_sdk.util
...
```

**(Valid for Helm based-operators only)** 

> The Helm operator packages as its base image was upgrade from Helm v2 to Helm v3. Note that cluster state for pre-existing CRs using Helm v2-based operators will be automatically migrated to Helm v3's new release storage format, and existing releases may be upgraded due to changes in Helm v3's label injection.

Reviewer: ```suggestion
The Helm operator packages and base image were upgraded from Helm v2 to Helm v3. Note that cluster state for pre-existing CRs using Helm v2-based operators will be automatically migrated to Helm v3's new release storage format, and existing releases may be upgraded due to changes in Helm v3's label injection.
```

If you are using any external helm v2 tooling with the your helm operator-managed releases, you will need to upgrade to the equivalent helm v3 tooling.

> For further detailed information see [CHANGELOG](https://github.com/operator-framework/operator-sdk/blob/master/CHANGELOG.md#v0141)

Reviewer: The way this is currently organized, this sentence appears to apply only to the Helm operator, when it is really applicable to the entire version's changeset.

```
github.com/operator-framework/operator-sdk/pkg/restmapper.DynamicRESTMapper
```

For:

```
sigs.k8s.io/controller-runtime/pkg/client/apiutil.DynamicRESTMapper
```

### Notable Changes

**(Valid for Ansible based-operators only)** 
- Extract custom Ansible module `k8s_status`, which is now provided by the `operator_sdk.util` Ansible collection. See [developer_guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/developer_guide.md#custom-resource-status-management) for new usage.
- Upgrade minimal Ansible version in the init projects from `2.6` to `2.9` for collections support. 

**(Valid for Helm based-operators only)** 
- Upgraded Helm operator packages and base image from Helm v2 to Helm v3. Cluster state for pre-existing CRs using Helm v2-based operators will be automatically migrated to Helm v3's new release storage format, and existing releases may be upgraded due to changes in Helm v3's label injection

For further detailed information see [CHANGELOG](https://github.com/operator-framework/operator-sdk/blob/master/CHANGELOG.md#v0141)

> 

Reviewer: Add another paragraph:

> If you are using any external helm v2 tooling with the your helm operator-managed releases, you will need to upgrade to the equivalent helm v3 tooling.

Author: Done solved.
